### DISCUSSION TITLE
[Proposal] Auto-Constructed Propoerties

### DISCUSSION THREAD

**Discussion by VBAndCs** [dotnet/csharplang#2768] [Language Ideas]
I suggest to use a Constructed attribute to auto initialize some properties and fields through the constructor. This will save a lot of coding in cases where there is no need to process the constructor params.. 
This will be helpful in many cases, such as dependency injection through constructors used a lot in controllers and PageModels in ASP.NET Core.

Example:
```C# 
class MyClass
{
    [Constructed] public int X { get; set; }
    [Constructed] public int Y { get; set; }
    [Constructed] public int Z { get; set; }

    public MyClass(string Msg)
    {
       //do something
    }
}
```

Roslyn can generate this code:
```C#
class MyClass
{
    public int X { get; set; }
    public int Y { get; set; }
    public int Z { get; set; }

    public MyClass(int x, int y, int z, string Msg)
    {
       X = x;
       Y = y;
       Z = z;
       //do something
    }
}
```

If we want a constructor that doesn't initialize the properties:
```C# 
  [NonConstructing] public MyClass(string Msg)
  {
       //do something
  }
```
Note that using attributes doesn't need to make any changes in the compiler. It can be done by Roslyn to lower the code directly. 
But @bandleader suggested two other alternatives to this syntax based on type script:
```C#
class MyClass
{
    public MyClass(public int X, public int Y , public int Z, string Msg)
    {
       //do something
    }
}
```

or:
```C#
class MyClass(public int X, public int Y , public int Z)
{
    public MyClass(string Msg)
    {
       //do something
    }
}
```

Which are more compact, but need to override the properties if we need to use write the get and set blocks.
The first syntax needs to define the properties repeatedly in all constructors. The second one overcomes that disadvantage, and if we use the [NonConstructing] attribute with some constructors, we will have the ability to define regular constructors as well.

So, the last syntax seems flexible and is also compact unless we have to override some properties to have a full body, where using the [Constructed] will be better.
In this case, I think we can have a mix of the first syntax and the third syntax to get the best of them.

**Comment by Joe4evr** [comment]
Look at Primary Constructors #2691

**Comment by VBAndCs** [comment]
@Joe4evr 
Nice. It is better to combine the two proposals like this:
1. add the property modifier in the definition of the Primary Constructors, and if not, public is the default one.
2. Use [NonConstructed] to exclude some constructors so they don't initialize the properties. In this case allow default values for the Primary Constructors.

Example:
```
class MyClass (private int X = 0, readonly int Y = 1, protected int Z = 2)
{
    [NonConstructed] public MyClass(string Msg)
    {
       //do something
    }
}
```

This will generate:
```C#
class MyClass 
{
  private int X {get; set;}
  public int Y {get;}
  protected int Z {get; set;}

    public MyClass(int x, int y, int z)
    {
       X = x;
       Y = y;
       Z = z;
    }

   public MyClass(string Msg)
    {
       X = 0;
       Y = 1;
       Z = 2;
       //do something
    }
}
```

**Comment by Thaina** [comment]
Are you trying to make `record` ? #39 

Also #2721

**Comment by VBAndCs** [comment]
@Thaina
No. I am trying to make regular mutable classes shorter.
